mongo/mongo.go

```diff
@@ -0,0 +1,362 @@
+
+	_, err := db.db.Collection(string(bucket)).Indexes().CreateOne(context.TODO(), index)
+	if err != nil {
+		return errors.Wrap(err, "error creating collection")
+	}
+
+	return nil
+}
+
+// DeleteTable deletes a root or embedded bucket. Returns an error if the
+// bucket cannot be found or if the key represents a non-bucket value.
+func (db *DB) deleteTable(bucket []byte, ctx context.Context) error {
+	if !collectionExists(db.db, string(bucket)) {
+		return errors.Wrapf(database.ErrNotFound, "bucket %s not found", bucket)
+	}
+
+	if err := db.db.Collection(string(bucket)).Drop(ctx); err != nil {
+		return errors.Wrapf(err, "error dropping collection: %s", bucket)
+	}
+	return nil
+}
+
+func (db *DB) get(bucket, key []byte, ctx context.Context) (ret []byte, err error) {
+	filter := createFilter("key", key)
+	res := db.db.Collection(string(bucket)).FindOne(ctx, filter)
+
+	if err := res.Err(); err != nil {
+		return nil, errors.Wrapf(database.ErrNotFound, "%s/%s", bucket, key)
+	}
+
+	result := Tuple{}
+	if err := res.Decode(&result); err != nil {
+		return nil, errors.Wrap(err, "error decoding value")
+	}
+
+	return []byte(result.Value), nil
+}
+
+func (db *DB) set(bucket, key, value []byte, ctx context.Context) error {
+	filter := createFilter("key", key)
+	update := createUpdate(value)
+	opts := options.Update().SetUpsert(true)
+
+	_, err := db.db.Collection(string(bucket)).UpdateOne(ctx, filter, update, opts)
+	if err != nil {
+		return errors.Wrapf(err, "error setting value %s/%s", bucket, key)
+	}
+	return nil
+}
+
+// List returns the full list of entries in a bucket.
+func (db *DB) list(bucket []byte, ctx context.Context) ([]*database.Entry, error) {
+	if !collectionExists(db.db, string(bucket)) {
+		return nil, errors.Wrapf(database.ErrNotFound, "bucket %s not found", bucket)
+	}
+
+	// match all
+	filter := bson.D{{}}
+
+	cursor, err := db.db.Collection(string(bucket)).Find(ctx, filter)
+	if err != nil {
+		return nil, errors.Wrap(err, "error listing values")
+	}
+	defer cursor.Close(context.Background())
+
+	if err = cursor.Err(); err != nil {
+		return nil, errors.Wrap(err, "error listing values")
+	}
+
+	var entries []*database.Entry
+
+	for cursor.Next(context.Background()) {
+		tuple := Tuple{}
+		cursor.Decode(&tuple)
+		entries = append(entries, &database.Entry{
+			Bucket: bucket,
+			Key:    []byte(tuple.Key),
+			Value:  []byte(tuple.Value),
+		})
+	}
+
+	if err = cursor.Err(); err != nil {
+		return nil, errors.Wrap(err, "error listing values")
+	}
+
+	return entries, nil
+}
+
+// Del deletes the value stored in the given bucket and key.
+func (db *DB) del(bucket, key []byte, ctx context.Context) error {
+	filter := createFilter("key", key)
+
+	mongoRes, err := db.db.Collection(string(bucket)).DeleteOne(ctx, filter)
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete %s/%s", bucket, key)
+	}
+
+	if mongoRes.DeletedCount == 0 {
+		return errors.Wrapf(err, "failed to delete: %s/%s. Value not found", bucket, key)
+	}
```

Comment:
Return a sentinel error that indicates the item wasn't found, so that this special case can be handled.